### PR TITLE
Guide.shapekey updates

### LINKED ISSUE #1366: making scatter plots with different shaped markers (issues)
# issue 1

```
plot(x=rand(4), y=rand(4), Geom.point, shape=[Shape.hexagon],
     Guide.xlabel("Re"),
     Guide.ylabel("Im")
)
```
automatically includes a legend, which I don't want. this seems an odd default, and how can I remove the legend?

image

# issue 2
this code results in no output
```
plot(layer(x=rand(4), y=rand(4), Geom.point, shape=[Shape.hexagon]),
     layer(x=rand(4), y=rand(4), Geom.point, shape=[Shape.cross]),
     Guide.xlabel("Re"),
     Guide.ylabel("Im")
)
```
how can I make two layers with different symbols outside (without building a `DataFrame`)?

# issue 3
the markers in the key seem to be a different size than inside the plot. is it possible to make them the same size using `Theme.point_size`?

image

# issue 4
even in the `DataFrame` paradigm, 

```
df = DataFrame(x=rand(3), y=rand(3), z=["a", "a", "b"])

plot(df, x=:x, y=:y, shape=:z, hipster_theme, shapes=[Shape.cross, Shape.circle])
```
how can I control which shapes are used, e.g. specify crosses and circles?

image

# issue 5
is there a manual key (legend) I could make while specifying the shapes of the data? That way I could get rid of the title on the legend above, which is "z".

### PR DESCRIPTION
**Pull request by Mattriks**
- [x] I've updated the documentation to reflect these changes
- [x] I've added and/or updated the unit tests
- [x] I've run the regression tests
- [x] I've built the docs and confirmed these changes don't cause new errors

### This PR:
- closes #1366 (see #1385 Roadmap)
- fixes triggering of shapekey (issue 1 in #1366, plot p1 below)
- fixes issue 6 in #1366 (plot p2)
- ~~shapekey integration with 1-length `size` and `color` vectors (plot p3)~~

### Example

```julia
x, y = 1:5, rand(5)
p1 = plot(x=x, y=y, shape=[Shape.square], size=[8pt])

p2 = plot(Theme(point_size=6pt, alphas=[0.6]),
    layer(x=0.5:4, y=rand(4), shape=["Gr1"]),
     layer(x=4*rand(4), y=rand(4), shape=["Gr2"]) )

p3 = plot(x=x, y=y, shape=x, size=[6pt], color=[colorant"orange"])
hstack(p1,p2,p3)
```
key_shape

### CODE REVIEW

**Comment by Mattriks** [comment]
The regression tests showed the following changes. With this PR, in testscripts:
1. "point_color_shape_size.jl", "point_shape_explicit.jl", and "shape_hvline.jl" the shapekey is not triggered (this is correct). Note that the examples in the first two testscripts are written in "old Gadfly style" (pre-shapekey) and could be updated in the future.

2. "key_columns.jl"  shows that the size aesthetic will override any default settings in `Theme`.  This maybe an issue in a plot that uses multiple aesthetics (e.g. distinct size and shape mappings).  Need to think more about plot p3 above (and if necessary add more to this PR).

**Comment by tlnagy** [comment]
This is great! I'm very excited for the propagation of a single color and shape through a layer. Makes multiple layer plots a lot easier. How hard would it be to autowrap non-array values? So you could do something like `color=colorant"orange"` instead of `color=[colorant"orange"]`, which makes more sense from a user's perspective.

**Comment by Mattriks** [comment]
placeholder

**Comment by Mattriks** [comment]
The regression tests now show changes to just the 3 testscripts listed in point 1 in my regression test post above. With this PR, in those 3 testscripts the shapekey doesn't trigger with `shape=[Shape.function]`. This is consistent with e.g.`color=[colorant"color"]` which doesn't trigger the colorkey.  

This PR is now ready to merge.

File: src/misc.jl [hunk 1]
```diff
@@ -338,6 +338,6 @@ function discretize_make_ia(values::CategoricalArray{T}, levels::Vector) where {
     return IndirectArray(index, convert(Vector{T},levels))
 end
 function discretize_make_ia(values::CategoricalArray{T}, levels::CategoricalVector{T}) where T
-    _levels = map!(t -> ismissing(t) ? t : get(t), Vector{T}(length(levels)), levels)
+    _levels = map!(t -> ismissing(t) ? t : get(t), Vector{T}(undef, length(levels)), levels)
```
Comment (bjarthur) [inline src/misc.jl:341]: good catch!  and great to see that codevec shows this line is now tested.

**Comment by bjarthur** [comment]
is it worth putting this in the gallery?

**Comment by Mattriks** [comment]
Done.  On Travis, for some reason the Codecov bit failed on Julia 1 now.

**Comment by Mattriks** [comment]
This is ready to merge.

**Pull request merged**